Firmware/mmu2.cpp

```diff
@@ -900,11 +908,50 @@ void MMU2::execute_extruder_sequence(const E_Step *sequence, uint8_t steps) {
 
 void MMU2::execute_load_to_nozzle_sequence() {
     planner_synchronize();
+    bool previous_ecool_state = load_unload_ecool_toggle(false);
     // Compensate for configurable Extra Loading Distance
     planner_set_current_position_E(planner_get_current_position_E() - (logic.ExtraLoadDistance() - MMU2_FILAMENT_SENSOR_POSITION));
     execute_extruder_sequence(load_to_nozzle_sequence, sizeof(load_to_nozzle_sequence) / sizeof(load_to_nozzle_sequence[0]));
+    (void)load_unload_ecool_toggle(previous_ecool_state);
 }
 
+bool MMU2::load_unload_ecool_toggle(bool enable){
+    // Check to see if E-cool mode is enabled and if so disable it during load sequence due to
+    // incompatibility with the MMU2 load sequence's many small moves.
+    // Fix for #3497.
+    if (enable == false) {
+        #ifdef TMC2130
+            if (farm_mode 
+            || (( eeprom_read_byte((uint8_t *)EEPROM_ECOOL_ENABLE) == EEPROM_ECOOL_MAGIC_NUMBER ) 
+                && ( eeprom_read_byte((uint8_t *)EEPROM_EXPERIMENTAL_VISIBILITY) == 1 ))
+            ){
+                st_synchronize();
+                if (tmc2130_wait_standstill_xy(1000)) {}
+                cli();
+                tmc2130_init(TMCInitParams(false)); 
+                st_reset_timer();
+                sei();
+                // Return that E-cool had previously been enabled.
+                return true;
+            }
+            else { // Return that E-cool had previously been disabled.
+                return false;
+            }
+        }
+        else { // Re-enable e-cool after it had been disabled.
+            st_synchronize();
+            if (tmc2130_wait_standstill_xy(1000)) {}
+            cli();
+            tmc2130_init(TMCInitParams(true)); 
+            st_reset_timer();
+            sei();
+        #endif //TMC2130
+    }
```

Comment:
I'll modify that slightly to always return false if TMC2130 is not defined, but otherwise that would be great!

Reply:
Haha, nevermind about that last comment, it was already set up for that.

Reply:
OK, so I've implemented the following: 
```C++
bool MMU2::load_unload_ecool_toggle(bool enable){
    // Check to see if E-cool mode is enabled and if so disable it during load sequence due to
    // incompatibility with the MMU2 load sequence's many small moves.
    // Fix for #3497.
    #ifdef TMC2130
        if (farm_mode
            || (( eeprom_read_byte((uint8_t *)EEPROM_ECOOL_ENABLE) == EEPROM_ECOOL_MAGIC_NUMBER )
            && ( eeprom_read_byte((uint8_t *)EEPROM_EXPERIMENTAL_VISIBILITY) == 1 )))
        {
            st_synchronize();
            if (tmc2130_wait_standstill_xy(1000)) {}
            cli();
            tmc2130_init(TMCInitParams(enable));
            st_reset_timer();
            sei();
            // Return that E-cool had previously been enabled.
            return true;
        } else { // Return that E-cool had previously been disabled, or always false if not TMC2130.
    #endif //TMC2130
            return false;
    #ifdef TMC2130
        }
    #endif //TMC2130
}
```

I'm not certain how to deal with the unused parameter for non-TMC2130 platforms, though.

Reply:
Also changing ```
if (farm_mode
    || (( eeprom_read_byte((uint8_t *)EEPROM_ECOOL_ENABLE) == EEPROM_ECOOL_MAGIC_NUMBER ) 
    && ( eeprom_read_byte((uint8_t *)EEPROM_EXPERIMENTAL_VISIBILITY) == 1 )))
``` to `if(FarmOrUserECool())` should save some memory too.

But we are trying to not include `ultralcd.h` into `mmu2.cpp` or `mmu2.h`.

I don't see any reason why `FarmOrUserECool()` should be in `ultralcd.cpp` 🤔 Maybe we can move it to `Marlin_main.cpp`

Reply:
>I'm not certain how to deal with the unused parameter for non-TMC2130 platforms, though.

IMO the code should only be run on TMC2130 printers. the `#ifdef` /`#endif` should otherwise prevent the code from being included in the firmware image.

Reply:
That was my initial thought, but it being in `ultralcd.cpp` did stop me. I'd ideally want to also move the E-cool enable/disable there also, as I think it's implemented in `ultralcd.cpp` also, then things might get smaller also without the duplicate toggling logic.

Reply:
I agree, this shouldn't be in the MMU2 class since this is technically on the printer side and not the MMU. Even so it seems strange to me to have this in `ultralcd.cpp` since this function is not UI related. Maybe part of this can be in `tmc2130.cpp` but its not clear to me at the moment if that's a clean solution.

Reply:
Do you know if there's a way to see if the currently running TMC2130 parameters are different from the setting in the EEPROM? I think I see a way forward that should be hopefully good if so.